### PR TITLE
Provide alternative to wrapping for FocusTrap.

### PR DESCRIPTION
**Pull request by JohnC-80**
Hello and thank you very much for this great library!
An issue that I've seen come up a few times here and in our own use of HotKeys has been some disturbance of the wrapping element that FocusTrap uses. I've seen the solution/response feedback in [#40](https://github.com/chrisui/react-hotkeys/pull/40) and opted to go a slightly different route that will fall back to default wrapping behavior.

I've introduced the prop 'noWrapper' to FocusTrap. If this is set to 'true',  after mounting, FocusTrap will attempt to find its 'trap' using `ReactDOM.findDOMNode(this);`. If that isn't successful, it sets an internal `willWrap` variable that will trigger the default wrapping behavior for the next update cycle. 
If the element is successfully discovered, it will apply `tabIndex -1` if necessary and native focus/blur listeners, passing in the `onFocus`, `onBlur` handlers provided via the props. This happens via helper functions -`addFocusListener()` and `addBlurListener()` that return removal methods to clean up those listeners when the component dismounts.

I look forward to any review/feedback from the maintainer/community and aim to actively conform this for hopeful inclusion.

### CODE REVIEW

**Comment by MikeTaylor** [comment]
I'd be keen to see this PR integrated.

**Comment by arecvlohe** [comment]
Any movement on this?

**Comment by MikeTaylor** [comment]
It looks like @chrisui is out of the game -- he's not responding to anything.

As a result, we reluctantly forked this code-base. If you'd like to use our modified version, you are welcome: it's at https://github.com/folio-org/stripes-react-hotkeys

We'd prefer that these changes be accepted into the original repo, but that doesn't seem to be happening. We're not even getting PRs rejected, they just sit there. I hope @chrisui is OK.

**Comment by greena13** [comment]
Hey guys, I am helping out with getting this package back up to speed. 

This pull request seems to be the most important to merge in.

File: lib/FocusTrap.js [hunk 1]
```diff
@@ -1,31 +1,91 @@
 import PropTypes from 'prop-types';
 import React from 'react';
+import ReactDOM from 'react-dom';
+import {addFocusListener, addBlurListener} from './focusListeners';
 
 class FocusTrap extends React.Component {
+  constructor(props) {
+    super(props);
+
+    this.focusListener;
```
Comment (greena13) [inline lib/FocusTrap.js]: What's the purpose of these two lines? They are not assignments, but do they perform some sort of initialisation I am not aware of?

File: lib/FocusTrap.js [hunk 2]
```diff
@@ -1,31 +1,91 @@
...
+  setupNativeListeners() {
+    if (!this.focusListener) {
+      const {
+        onFocus, 
+        onBlur, 
+      } = this.props;
+
+      const elem = ReactDOM.findDOMNode(this);
+
+      if (elem && !this.focusListener) {
```
Comment (greena13) [inline lib/FocusTrap.js]: Is `!this.focusListener` not redundant, given the outer condition already checks for this?

File: lib/FocusTrap.js [hunk 3]
```diff
@@ -1,31 +1,91 @@
...
+        this.focusListener = addFocusListener(elem, onFocus);
+        this.blurListener = addBlurListener(elem, onBlur);
+      } else {
+        this.willWrap = true;
+      }
+    }
+  }
+
+  removeNativeListeners() {
+    if (this.focusListener) {
```
Comment (greena13) [inline lib/FocusTrap.js]: It may make sense to record whether we add a new `tabindex` value on mount, so that we can remove it again on dismount to completely clear up after ourselves - unless this is always taken care of by the act of unmounting, anyway?

File: lib/FocusTrap.js [hunk 4]
```diff
@@ -1,31 +1,91 @@
...
+import ReactDOM from 'react-dom';
+import {addFocusListener, addBlurListener} from './focusListeners';
 
 class FocusTrap extends React.Component {
+  constructor(props) {
+    super(props);
+
+    this.focusListener;
+    this.blurListener;
+    this.willWrap = !this.props.noWrapper;
```
Comment (greena13) [inline lib/FocusTrap.js]: Should `willWrap` not be a part of the component's `state` to trigger an immediate re-render in the event that it is set to `false` after the component has mounted? Otherwise, if I am interpreting the code and component lifecycle correctly, the component will need to wait for some change in props to re-render the component.

File: lib/FocusTrap.js [hunk 5]
```diff
@@ -1,31 +1,91 @@
...
+    this.setupNativeListeners = this.setupNativeListeners.bind(this);
+    this.removeNativeListeners = this.removeNativeListeners.bind(this);
+  }
+
   static propTypes = {
     onFocus: PropTypes.func,
     onBlur: PropTypes.func,
     component: PropTypes.any,
+    noWrapper: PropTypes.bool,
```
Comment (greena13) [inline lib/FocusTrap.js]: I am not a huge fan of this property name. Working with negatives can be harder and confusing. It is also a little obtuse - something like `wrapChildren` or `useFocusWrapper` is more direct and explanatory about the behaviour the user can expect. 

This will of course require complementing the logic throughout the component, and setting a default value of `true`.

**Comment by JohnC-80** [comment]
Bare with me on tackling these changes! It's been a little while since I submitted this... Thanks for taking maintainership of the library!

File: lib/FocusTrap.js (hunk 5 above)
Comment (JohnC-80) [reply lib/FocusTrap.js]: I understand the opposition to 'turning on an off switch' with the negatory prop name. Will work with updating it to use `useFocusWrapper`.

File: lib/FocusTrap.js (hunk 2 above)
Comment (JohnC-80) [reply lib/FocusTrap.js]: Yep. Will clean up!

File: lib/FocusTrap.js (hunk 4 above)
Comment (JohnC-80) [reply lib/FocusTrap.js]: For instances where the component attempts to `findDOMNode()` but fails - ok, will do.

File: lib/FocusTrap.js (hunk 3 above)
Comment (JohnC-80) [reply lib/FocusTrap.js]: I think that in order for a parent to unmount, the children would have to unmount as well, right? Is there a use case where <HotKeys> and <FocusTrap> would go away without doing an actual remount-from-scratch of the contained children? The code for this wouldn't be very complex, but I think we may need to wait for a specific need for this, unless you've got a specific one in mind.

Comment (greena13) [reply lib/FocusTrap.js]: No, I think you are right that without exception, this would be taken care of anyway by all children being unmounted. I'm happy to leave this out for now.

**Comment by JohnC-80** [comment]
So I've added the basic test to cover the `useFocusWrap` with the component being found... I did have to get a little adventurous there and use enzyme's `.getDOMNode()` in order to properly get the element that I could focus and that the native DOM listeners would respond to. This may have something to do with "simulating" focus in `FocusableElement`, but I'm not quite sure. I've run into some issues in the test setup (I think) when attempting to build a test case for the component NOT being found... I was thinking of just passing in a stateful component that would render null, effectively not giving you a DOM node that would be latch-able, but I'm running into some odd hurdles that I'll post more about later once I've put a little more time into it. Please have a look at things, @greena13 and let me know any more thoughts you have. Thanks!

**Comment by greena13** [comment]
The use of `.getDOMNode()` is fine (and preferable - I couldn't find such a method when I was writing the test suite).

The simulation of key events in the test suite is definitely a little hairy but it was the only way I could get it to function. None of enzyme's event simulation triggers the hotkey actions (because the package doesn't actually use React simulated events & event propagation to work). Unfortunately, it does come with some limitations and side effects that can be frustrating. 

Could you simply render 

```
<HotKeys keyMap={this.keyMap} handlers={handlers}>
  <input />
</HotKeys>
```

And assert that a `<div />` is present in the children? 

[Chai-Enzyme](https://github.com/producthunt/chai-enzyme) (already used in the test suite) should allow you to assert both the presence of the div and all the important attributes it should have.

**Comment by JohnC-80** [comment]
Yes, I've finished that case here: https://github.com/chrisui/react-hotkeys/pull/61/files#diff-5f13dc3d3d804a45a58a70134f5fe435 
It's the case where there's no DOM element in the children to attach to that I'm having difficulty with (or any circumstance where ReactDOM.findDOMNode() would return `null`  - the prime case for this in my mind is if a child stateful component returns null - something like 
```
import React from 'react';

class NullComponent extends React.Component {
  render() {
    return null;
  }
}

export default NullComponent;
```
Outside of this sort of instance, it seems challenging to not be able to get a DOMNode successfully (especially in more recent versions of react)
I'm attempting to include this in the test suite, but when I do I get an error: 
```

**Comment by greena13** [comment]
Ah, I see. I misunderstood you.

I agree with your thinking and off the top of my head, I'm not sure what may be causing that error. If you would like some help trying to diagnose, you could share or push the code that generates that exception and I could attempt to assist. :) 

The mention of not being able to render in a worker thread reminds me of the [new feature added in React 16](https://reactjs.org/blog/2017/09/26/react-v16.0.html#new-core-architecture).

**Comment by JohnC-80** [comment]
@greena13 I've added the test for when the DOM node isn't found as well as the supplementary NullComponent - The test still fails in the testing framework with the odd error I previously mentioned. Apologies for the delay.

**Comment by JohnC-80** [comment]
Perhaps the focusTrap actually needs non-empty/non-null children in it to work when it's left up to do the wrapping?
In a browser, using <NullComponent>, I can inspect, see the `<div tabindex="-1">` and in ReactDevTools I can see that it does have the `onFocus` and `onBlur` props passed in - of course, with a tabindex of -1 it isn't keyboard focusable or testable... and there are no focusable children, just a `React-empty` comment node.

**Comment by greena13** [comment]
Hey John, apologies for letting this one slide through the cracks. I'm just now finding time to return to maintaining this project. 

Do you still have the time and the will to continue with this, or would you like me to take over and try and find a suitable resolution?

**Comment by greena13** [comment]
I'm wondering if rather than define a new prop `noWrap`, we can re-use the existing [`component` prop](https://github.com/greena13/react-hotkeys/blob/master/lib/FocusTrap.js#L18), which is used to indicate the type of component to use to wrap `HotKeys`'s children.

Perhaps we could use a `false`y value for `component` to not render a focus trap. That does have the drawback however, if a suitable DOM element cannot be found to serve as the focus trap, there is no way of specifying a fallback element other than (the default)`div`.

**Comment by JohnC-80** [comment]
Seems like the falseyness of `component` would get a bit weird and we would have to get specific on falseyness - where `undefined` would get you the default  `<div>`, but `false` would get you the non-wrapped behavior... I think I'd still prefer the `noWrap` or `useWrapper={false}` since they're a bit more straightforward in what they cause to happen. 
The DOM dependency of the wrapper tripped me up in the testing (more specifically, how to test on something that's empty when focus can't even happen there.) Updates in React16 have complicated things a little bit more - if using a `<React.Fragment>` as a child would result in only the first element in the fragment's children being returned in a call to `ReactDOM.findDOMNode()` which would be inadequate to properly catch events - so an exception would have to be devised there.
I've been deeply engaged in other portions of the project I'm tasked with outside of the Hotkeys realm - but a `noWrap` prop has worked quite well there thus far, and there aren't many instances of fragments or arrays that need hotkeys without obvious parent wrappers (at least not yet.)

**Comment by greena13** [comment]
I think you're right about the about (mis)using falsey values of `component`.

I like the idea of a non-negative prop name as well (`useWrapper` with a default value of `true` rather than a prop `noWrap` with a default value of false). Perhaps it could be shortened to just `wrap`.

If I understand you correctly, the problem is testing the case that `HotKeys` is used with `wrap={false}` AND nothing is returned by `ReactDOM.findDOMNode()`. When you attempt to write a test for this case (using `NullComponent`), you got the error above about `dangerouslyReplaceNodeWithMarkup`.

Do we need the implicit fallback behaviour of rendering the nominated `component` if `ReactDOM.findDOMNode()` returns `null`? Perhaps it is better to simply display a warning if this happens and let the library user resolve the issue.  

I'm going to see if I can get my head around the [current behaviour of ReactDOM.findDOMNode()](https://github.com/facebook/react/blob/master/packages/react-dom/src/client/ReactDOM.js#L1121)  to fully comprehend its nuances.

**Comment by greena13** [comment]
Ok, if I am reading the source code correctly, `ReactDOM.findDOMNode()` only ever considers the react component passed to it, and its children. 

So I think in our case, we have two situations where it could return `null`. When `HotKeys`'s children are:

* `null`
* An empty array
* A component that returns `null` or an empty array
* An array of components that return `null` or an empty array

I think the [expected behaviour](https://en.wikipedia.org/wiki/Principle_of_least_astonishment) in each of those cases would be for `HotKeys` to quietly not attach any listeners or `tabIndex={-1}` to anything.

The trouble with that is that when the children change (the null component's state changes and it returns something) `HotKeys` is not going to detect the change so the hot keys are not going to be active until something updates the state or props of the `HotKeys` component. 

I don't currently know how to deal with the `Fragment` problem, either. Is it well documented or is through your own testing, @JohnC-80, that you know `ReactDOM.findDOMNode()`, when passed a `Fragment`, will return the first element in the array it contains?

**Comment by heyjohnnyc** [comment]
That behavior is documented here https://reactjs.org/docs/react-dom.html#finddomnode

**Comment by JohnC-80** [comment]
FindDOMNode's behavior with Fragments is documented here: https://reactjs.org/docs/react-dom.html#finddomnode

**Comment by JohnC-80** [comment]
@greena13 it's been a while since I've originally posted this and there's been a number of changes in the library since. A conversation with a colleague sparked an idea for handling the situation with Fragments - or virtually any kind of container/shape where you'd want to apply HotKeys... instead of only a boolean, have the 'focused' prop also accept a function - so that it could be checked at handle-time to see whether or not to handle the shortcut. This inverts control but leaves the situation open to any logic a developer may need to use in order to 'quarantine' their unwrapped keyable set.

**Comment by greena13** [comment]
Closing, as I think this was superseded by a [similar PR](https://github.com/greena13/react-hotkeys/pull/125) that was also closed.

Thanks again for the effort you put into this over the months.
